comp/core/agenttelemetry/agenttelemetryimpl/config.go

```diff
@@ -0,0 +1,526 @@
+		if err := compileMetric(p, &p.Metric.Metrics[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Compile status section
+func compileStatus(p *Profile) error {
+	// No status section - nothing to do
+	if p.Status == nil {
+		return nil
+	}
+
+	// Validate template (optional with default "basic". "none" is also supported)
+	if len(p.Status.Template) == 0 {
+		p.Status.Template = "basic"
+	} else if p.Status.Template != "basic" && p.Status.Template != "none" {
+		return fmt.Errorf("profile '%s' template attribute can have 'basic' or 'none' value but %s is provided", p.Name, p.Status.Template)
+	}
+
+	// Compile status extra
+	var err error
+	p.statusExtraBuilder, err = compileJBuilders(p.Status.Extra)
+	if err != nil {
+		return fmt.Errorf("failed to compile 'extra' attribute for profile '%s'. Error: %w", p.Name, err)
+	}
+
+	return nil
+}
+
+// Compile profile
+func compileProfile(p *Profile) error {
+	// Profile requires name
+	if len(p.Name) == 0 {
+		return fmt.Errorf("profile requires 'name' attribute to be specified")
+	}
+
+	if err := compileMetrics(p); err != nil {
+		return err
+	}
+
+	if err := compileStatus(p); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Compile schedules
+func compileSchedules(cfg *Config) error {
+	cfg.schedule = make(map[Schedule][]*Profile)
+
+	for i := 0; i < len(cfg.Profiles); i++ {
+		p := &cfg.Profiles[i]
+
+		// Setup default schedule if it is not specified partially or at all
+		if p.Schedule == nil {
+			p.Schedule = &Schedule{
+				StartAfter: defaultSheduleStartAfter,
+				Iterations: 0,
+				Period:     defaultShedulePeriod,
+			}
+		} else {
+			// Validate StartAfter (optional)
+			if p.Schedule.StartAfter == 0 {
+				p.Schedule.StartAfter = defaultSheduleStartAfter
+			}
+
+			// Validate Period (optional)
+			if p.Schedule.Period == 0 {
+				p.Schedule.Period = defaultShedulePeriod
+			}
+		}
+
+		// Aggregate schedules (one schedule correspond to one or more profiles)
+		if pp, ok := cfg.schedule[*p.Schedule]; ok {
+			pp = append(pp, p)
+			cfg.schedule[*p.Schedule] = pp
+		} else {
+			pp = []*Profile{p}
+			cfg.schedule[*p.Schedule] = pp
+		}
+	}
+
+	return nil
+}
+
+// Compile agent telemetry config
+func compileConfig(cfg *Config) error {
+	for i := 0; i < len(cfg.Profiles); i++ {
+		err := compileProfile(&cfg.Profiles[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := compileSchedules(cfg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Parse agent telemetry config
+func parseConfig(cfg config.Component) (*Config, error) {
+	// Is it enabled?
+	if !cfg.GetBool("agent_telemetry.enabled") {
+		return &Config{
+			Enabled: false,
+		}, nil
+	}
+
+	var atCfg Config
+
+	// Parse agent telemetry config
+	atCfgMap := cfg.GetStringMap("agent_telemetry")
+	if len(atCfgMap) > 0 {
+		// Reconvert to string and back to object.
+		// Config.UnmarshalKey() is better but it did not work in some cases
+		atCfgBytes, err := yaml.Marshal(atCfgMap)
```

Comment:
I'm not entire sure how are config parsing logic works 100%. Does the first pass to the config parsed nested configuration? Mostly asking to make sure we avoid doing any extra `yaml.Marshal`

Reply:
I have hard time to extract more complex configuration hierarchy. One method work but it seems to have limitations. My double marshal/unmarshall is a work around for that since I do want to have unmarshal into dedicated structure. I did left a comment of that sort.

Reply:
Closing per OOTB discussion